core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala

```diff
@@ -161,22 +162,26 @@ class TransactionStateManager(brokerId: Int,
             maxBatchSize
           )
 
-          partitionCacheEntry.metadataPerTransactionalId.foreachWhile { (transactionalId, txnMetadata) =>
+          var breakIteration = false
```

Comment:
I think I may miss something here, could you explain why we'd need both the caller/callee doing this while loop with the break iteration flag? It seems to me that we have multiple reasons to break early: 1) we've reached the record limit, 2) the partition is already offline, and we want to treat them differently and hence the both while loops, right?

But I cannot see exactly where we may have case 2)?

Reply:
We've changed the logic to add check the log configuration to get to the batch size. If the partition is offline, then `ReplicaManager.getLogConfig` will return `None`. I have a test case which shows this path.

Reply:
In that case, I feel it may actually get cleaner to inline `collectExpiredTransactionalIds` into the caller, and hence to get just one while loop / flag, we can still distinguish the case where the log is offline and hence we should not proceed v.s. the batch is full, we should write once and proceed. WDYT?

Reply:
I tried to consolidate this logic into a single loop, which I think is what you are suggesting. Let me know if the latest commit is what you had in mind.